File: scripts/js/tokko_backfill_existing_leads.js
const fs = require('fs');
const path = require('path');
const { Client } = require('/home/deploy/atendechat/backend/node_modules/pg');

(async () => {
  const rtPath = '/home/deploy/atendechat/backend/runtime-settings.json';
  const rt = JSON.parse(fs.readFileSync(rtPath, 'utf8'));
  if (!rt.tokkoEnabled) throw new Error('tokkoEnabled=false');

  const envRaw = fs.readFileSync('/home/deploy/atendechat/backend/.env', 'utf8');
  const env = {};
  for (const line of envRaw.split(/\r?\n/)) {
    if (!line || line.trim().startsWith('#') || !line.includes('=')) continue;
    const i = line.indexOf('=');
    env[line.slice(0, i).trim()] = line.slice(i + 1).trim();
  }

  const apiKey = String(rt.tokkoApiKey || env.TOKKO_API_KEY || '').trim();
  if (!apiKey) throw new Error('missing tokkoApiKey');

  const base = String(rt.tokkoBaseUrl || 'https://www.tokkobroker.com/api/v1').replace(/\/$/, '');
  const leadsPath = String(rt.tokkoLeadsPath || '/webcontact/').startsWith('/') ? String(rt.tokkoLeadsPath || '/webcontact/') : `/${String(rt.tokkoLeadsPath || 'webcontact/')}`;
  const endpoint = `${base}${leadsPath}`;
  const url = new URL(endpoint);
  url.searchParams.set('key', apiKey);

  const db = new Client({
    host: env.DB_HOST || '127.0.0.1',
    port: Number(env.DB_PORT || 5432),
    user: env.DB_USER,
    password: env.DB_PASS,
    database: env.DB_NAME,
  });
  await db.connect();

  await db.query(`INSERT INTO tags (name, color, "createdAt", "updatedAt") VALUES ('enviado_tokko', '#0EA5E9', NOW(), NOW()) ON CONFLICT (name) DO NOTHING`);
  const tagRes = await db.query(`SELECT id FROM tags WHERE name='enviado_tokko' LIMIT 1`);
  const tagId = Number(tagRes.rows[0].id);

  const q = await db.query(`
    SELECT c.id, COALESCE(c.name,'') AS name, COALESCE(c.number,'') AS number, COALESCE(c.email,'') AS email, COALESCE(c.source,'') AS source
    FROM contacts c
    WHERE c."companyId" = 1
      AND COALESCE(c."isGroup", false) = false
      AND COALESCE(regexp_replace(c.number, '\\D', '', 'g'),'') <> ''
      AND NOT EXISTS (
        SELECT 1 FROM contact_tags ct
        WHERE ct."contactId" = c.id AND ct."tagId" = $1
      )
    ORDER BY c.id ASC
  `, [tagId]);

  let sent_ok = 0, failed = 0, skipped = 0;
  for (const row of q.rows) {
    const phone = String(row.number || '').replace(/\D/g, '');
    if (!phone) { skipped++; continue; }

    const payload = {
      name: String(row.name || phone).slice(0, 120),
      email: String(row.email || '').slice(0, 180),
      phone,
      text: 'Backfill automático desde Charlott CRM',
      source: String(row.source || 'backfill-existing-leads'),
      tags: ['Lead_Calificado', 'Bot']
    };

    try {
      const resp = await fetch(url.toString(), {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(payload)
      });
      if (resp.ok) {
        sent_ok++;
        await db.query(`INSERT INTO contact_tags ("contactId", "tagId", "createdAt", "updatedAt") VALUES ($1,$2,NOW(),NOW()) ON CONFLICT DO NOTHING`, [Number(row.id), tagId]);
      } else {
        failed++;
      }
    } catch {
      failed++;
    }
  }

  console.log(JSON.stringify({ candidates: q.rows.length, sent_ok, failed, skipped, tag_id: tagId, endpoint }));
  await db.end();
})().catch((e) => { console.error(String(e?.message || e)); process.exit(1); });

File: scripts/python/tokko/tokko_backfill_existing_leads.py
import json
import re
from pathlib import Path
from urllib.parse import urlencode

import psycopg2
import requests

rt_path = Path('/home/deploy/atendechat/backend/runtime-settings.json')
rt = json.loads(rt_path.read_text(encoding='utf-8')) if rt_path.exists() else {}

if not rt.get('tokkoEnabled'):
    raise SystemExit('ERROR: tokkoEnabled=false')

api_key = str(rt.get('tokkoApiKey') or '').strip()
base = str(rt.get('tokkoBaseUrl') or 'https://www.tokkobroker.com/api/v1').rstrip('/')
leads_path = str(rt.get('tokkoLeadsPath') or '/webcontact/')
if not leads_path.startswith('/'):
    leads_path = '/' + leads_path
if not api_key:
    raise SystemExit('ERROR: missing tokkoApiKey')

url = f"{base}{leads_path}?" + urlencode({'key': api_key})

env = {}
for line in Path('/home/deploy/atendechat/backend/.env').read_text(encoding='utf-8').splitlines():
    if '=' in line and not line.strip().startswith('#'):
        k, v = line.split('=', 1)
        env[k.strip()] = v.strip()

conn = psycopg2.connect(
    host=env.get('DB_HOST', '127.0.0.1'),
    port=int(env.get('DB_PORT', '5432')),
    user=env.get('DB_USER'),
    password=env.get('DB_PASS'),
    dbname=env.get('DB_NAME')
)
conn.autocommit = True
cur = conn.cursor()

cur.execute(
    '''
    INSERT INTO tags (name, color, "createdAt", "updatedAt")
    VALUES ('enviado_tokko', '#0EA5E9', NOW(), NOW())
    ON CONFLICT (name) DO NOTHING
    '''
)
cur.execute("SELECT id FROM tags WHERE name='enviado_tokko' LIMIT 1")
tag_id = cur.fetchone()[0]

cur.execute(
    '''
    SELECT c.id, COALESCE(c.name,''), COALESCE(c.number,''), COALESCE(c.email,''), COALESCE(c.source,'')
    FROM contacts c
    WHERE c."companyId" = 1
      AND COALESCE(c.isGroup, false) = false
      AND COALESCE(regexp_replace(c.number, '\\D', '', 'g'),'') <> ''
      AND NOT EXISTS (
        SELECT 1 FROM contact_tags ct
        WHERE ct."contactId" = c.id AND ct."tagId" = %s
      )
    ORDER BY c.id ASC
    ''',
    (tag_id,)
)
rows = cur.fetchall()

ok = 0
fail = 0
skipped = 0

for cid, name, number, email, source in rows:
    phone = re.sub(r'\D', '', number or '')
    if not phone:
        skipped += 1
        continue

    payload = {
        'name': (name or phone)[:120],
        'email': (email or '')[:180],
        'phone': phone,
        'text': 'Backfill automático desde Charlott CRM',
        'source': source or 'backfill-existing-leads',
        'tags': ['Lead_Calificado', 'Bot']
    }

    try:
        r = requests.post(url, json=payload, timeout=20)
        if 200 <= r.status_code < 300:
            ok += 1
            cur.execute(
                '''
                INSERT INTO contact_tags ("contactId", "tagId", "createdAt", "updatedAt")
                VALUES (%s, %s, NOW(), NOW())
                ON CONFLICT DO NOTHING
                ''',
                (cid, tag_id)
            )
        else:
            fail += 1
    except Exception:
        fail += 1

print(json.dumps({
    'candidates': len(rows),
    'sent_ok': ok,
    'failed': fail,
    'skipped': skipped,
    'tag_id': tag_id,
    'endpoint': f"{base}{leads_path}"
}, ensure_ascii=False))

cur.close()
conn.close()
